File: packages/telegram_bot/src/telegram_bot/reporting.py
import logging
from dataclasses import dataclass

import httpx

from telegram_bot.extractor import ExtractionResult, extraction_to_log
from telegram_bot.transit import TransitData, resolve_line_variant

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class ReportClient:
    client: httpx.AsyncClient
    headers: dict[str, str]


def build_report_payload(
    *,
    station_id: str,
    line_id: str | None,
    direction_id: str | None,
) -> dict[str, object]:
    payload: dict[str, object] = {
        'stationId': station_id,
        'source': 'telegram',
    }
    if line_id is not None:
        payload['lineId'] = line_id
    if direction_id is not None:
        payload['directionId'] = direction_id
    return payload


def report_identifiers(transit: TransitData, extraction: ExtractionResult) -> tuple[str, str | None, str | None] | None:
    if extraction.is_empty:
        logger.info('LLM returned no inspector report for this message')
        return None

    if extraction.station_id is None:
        # Backend requires at least one identifier and our pipeline keys off the station.
        logger.info('Extraction lacked stationId; skipping (%s)', extraction_to_log(extraction))
        return None

    line_id = (
        resolve_line_variant(transit, extraction.line_name, extraction.station_id)
        if extraction.line_name is not None
        else None
    )
    return extraction.station_id, line_id, extraction.direction_id


async def submit_report(
    reports: ReportClient,
    *,
    station_id: str,
    line_id: str | None,
    direction_id: str | None,
) -> bool:
    payload = build_report_payload(station_id=station_id, line_id=line_id, direction_id=direction_id)

    try:
        response = await reports.client.post('/v0/reports', json=payload, headers=reports.headers)
    except httpx.HTTPError:
        logger.exception('POST /v0/reports failed')
        return False

    if response.is_success:
        return True

    logger.error('POST /v0/reports returned %s: %s', response.status_code, response.text)
    return False


async def close_report_client(reports: ReportClient) -> None:
    await reports.client.aclose()

File: packages/telegram-worker/src/reporting.ts
import type { CityConfig } from '@freifahren/cities'

import type { TrustedReportGate } from './report-gate-contract'
import type { TransitIndex } from './types'
import { type ExtractionResult, extractionToLog, isEmpty } from './extractor'
import { resolveLineVariant } from './transit'

export interface ReportPayload {
    stationId: string
    source: 'telegram'
    lineId?: string
    directionId?: string
}

export interface ReportIdentifiers {
    stationId: string
    lineId: string | null
    directionId: string | null
}

export function buildReportPayload(ids: ReportIdentifiers): ReportPayload {
    const payload: ReportPayload = { stationId: ids.stationId, source: 'telegram' }
    if (ids.lineId !== null) {
        payload.lineId = ids.lineId
    }
    if (ids.directionId !== null) {
        payload.directionId = ids.directionId
    }
    return payload
}

export function reportIdentifiers(index: TransitIndex, extraction: ExtractionResult): ReportIdentifiers | null {
    if (isEmpty(extraction)) {
        console.info('LLM returned no inspector report for this message')
        return null
    }
    if (extraction.stationId === null) {
        // Backend requires at least one identifier and our pipeline keys off the station.
        console.info('Extraction lacked stationId; skipping', extractionToLog(extraction))
        return null
    }

    const lineId =
        extraction.lineName !== null ? resolveLineVariant(index, extraction.lineName, extraction.stationId) : null

    return { stationId: extraction.stationId, lineId, directionId: extraction.directionId }
}

// Throws on a rejected RPC result; the caller reports it (no retry).
export async function postReport(
    reportGate: TrustedReportGate,
    ids: ReportIdentifiers,
    city: CityConfig
): Promise<void> {
    const result = await reportGate.intake({
        city: {
            slug: city.slug,
            publicAppUrl: city.publicAppUrl,
            dbBinding: city.dbBinding,
            telegramChatId: city.community.telegramChatId ?? null,
            reporting: city.reporting,
        },
        report: buildReportPayload(ids),
    })
    if (!result.ok) {
        throw new Error(`Trusted report intake failed: ${result.error.internalCode}`)
    }
}
